bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py

```diff
@@ -0,0 +1,278 @@
+                'int': nb.int32,
+                'unsigned int': nb.uint32,
+                'long': nb.int64,
+                'unsigned long': nb.uint64,
+                'bool': nb.boolean
+                }
+        if t in typemap:
+            return typemap[t]
+        raise Exception(
+                'Type {} is not supported for jitting with numba. Valid fundamental types and RVecs thereof are {}'.format(
+                    t, list(typemap.keys())))
+
+    def get_c_signature(input_types, return_type):
+        '''
+        Get C signature as numba type objects from C++ typenames
+        '''
+        c_input_types = []
+        for t in input_types:
+            if 'RVec' in t:
+                c_input_types += [nb.types.CPointer(get_numba_type(get_inner_type(t))), nb.int32]
+            else:
+                c_input_types.append(get_numba_type(t))
+        if 'RVec' in return_type:
+            # See the Python wrapper for the reason using these types
+            c_return_type = nb.void
+            c_input_types += [nb.types.CPointer(nb.int64), nb.types.CPointer(nb.int64), nb.int32]
+        else:
+            c_return_type = get_numba_type(return_type)
+        return c_return_type, c_input_types
+
+    def get_numba_signature(input_types, return_type):
+        '''
+        Get numba signature as numba type objects from C++ typenames
+        '''
+        nb_input_types = []
+        for t in input_types:
+            if 'RVec' in t:
+                nb_input_types.append(get_numba_type(get_inner_type(t))[:])
+            else:
+                nb_input_types.append(get_numba_type(t))
+        if 'RVec' in return_type:
+            nb_return_type = get_numba_type(get_inner_type(return_type))[:]
+        else:
+            nb_return_type = get_numba_type(return_type)
+        return nb_return_type, nb_input_types
+
+    # Define inner decorator without arguments
+    def inner(func, input_types=input_types, return_type=return_type, name=name):
+
+        # Jit the given Python callable with numba
+        nb_return_type, nb_input_types = get_numba_signature(input_types, return_type)
+        try:
+            nbjit = nb.jit(nb_return_type(*nb_input_types), nopython=True, inline='always')(func)
+        except:
+            raise Exception('Failed to jit Python callable {} with numba.jit'.format(func))
+        func.__numba_jit__ = nbjit
+
+        # Create Python wrapper with C++ friendly signature
+
+        # Define signature
+        pywrapper_signature = [
+                'ptr_{0}, size_{0}'.format(i) if 'RVec' in t else 'x_{}'.format(i) \
+                        for i, t in enumerate(input_types)]
+        if 'RVec' in return_type:
+            # If we return an RVec, we return via pointer the pointer of the allocated data,
+            # the size in elements. In addition, we provide the size of the datatype in bytes.
+            pywrapper_signature += ['ptrptr_r, ptrsize_r', 'dtypesize_r']
```

Comment:
Could the logic be simplified a bit by removing the `dtypesize_r` argument from the signature and just use the value we know it has (because numba types are always strictly "sized")?

Reply:
I get the type via a `sizeof` call from C++, but you are also right, we could go through numba and get the size of the type from them. I check whether this is easily possible because I would like to prevent another map!

Reply:
i'd say even another map would be a simplification w.r.t. generating code that conditionally has the extra argument and then generating code that conditionally passes the argument after calling sizeof(...) from jitted C++ :sweat_smile:

Reply:
but you have to maintain the map, not the logic :) though, i understand your point

Reply:
true, up to you

Reply:
I tried to get in a `numba.float32.bitwidth` or similar. Unfortuantley numba doesn't recognize the type / the member function in the jitting with `nopython=True`. So i'll stick to this solution :/

Reply:
aaah `bitwidth`, well done finding that one, I couldn't when I looked for it! You should not call `numba.float32.bitwidth` from the jitted function -- you can just inject the actual value in the generated code. that simplifies the logic quite a bit

Reply:
ah i'm stupid :see_no_evil:

Reply:
Thanks, works :dancing_men:

Reply:
wo-hoo! :smile: